File: cobalt/frontend/src/services/api/ws/servers.ts
/*
 * Copyright (C) 2026 Artorias
 * Author: Artorias
 * Repository: https://github.com/artorias-developer/cobalt
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import type { IWsClient, IWsServersApiService } from "@/contracts"
import { ServersEventEnum } from "@/types"

/**
 * WebSocket service for servers events.
 */
export class WsServersApiService implements IWsServersApiService {
  private readonly client: IWsClient

  constructor(client: IWsClient) {
    this.client = client
  }

  /**
   * Subscribes to servers states and registers a handler.
   *
   * Parameters:
   * - handler: Callback function for incoming server states.
   *
   * Returns:
   * - void.
   */
  subscribeStates(handler: (states: any) => void): void {
    this.client.listen(ServersEventEnum.SERVER_STATE, handler)
    this.client.subscribe(ServersEventEnum.SUBSCRIBE_STATES)
  }

  /**
   * Unsubscribes from servers states and removes the handler.
   *
   * Parameters:
   * - handler: Callback function to remove.
   *
   * Returns:
   * - void.
   */
  unsubscribeStates(handler: (states: any) => void): void {
    this.client.unlisten(ServersEventEnum.SERVER_STATE, handler)
    this.client.unsubscribe(ServersEventEnum.UNSUBSCRIBE_STATES)
  }
}

File: cobalt/backend/presentation/contracts/ws/events/servers.py
#  Copyright (C) 2026 Artorias
#  Author: Artorias
#  Repository: https://github.com/artorias-developer/cobalt
#  SPDX-License-Identifier: AGPL-3.0-or-later

from abc import abstractmethod, ABC
from typing import Any

from presentation.contracts.ws.routers import AbstractWsRouter


class AbstractWsServersEvents(AbstractWsRouter, ABC):
    """
    Abstract WebSockets events for status-related operations.
    """

    @abstractmethod
    async def subscribe_states(self, *args: Any, **kwargs: Any) -> None:
        """
        Subscribes to servers states.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - None.
        """
        ...

    @abstractmethod
    async def unsubscribe_states(self, *args: Any, **kwargs: Any) -> None:
        """
        Unsubscribes from servers states.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - None.
        """
        ...
